Activate November 15th 2018 upgrade by height

sickpig opened the pull request:
This is also add the helper that have to be used to activate consensus rule changes that come with May 15th, 2019 network upgrade.

Almost done, I only need to fix `excessive.py` and I could use some help on that :P.

sickpig committed:
Fix a bunch of comments that refers to UAHF improperly

UAHF was the first of series of protocol upgrades that defines
the current set of BCH consensus rules. Sometimes in the comment
it was used to identify the BCH chain, even if this is partially
true (w/o UAHF there wouldn't be a BCH chain), it is not correct.

sickpig committed:
Removed unsued UAHF fork helper

sickpig committed:
Start to introduce changes need for the May 15th 2019

Add Nov 15th 2018 block height activation
Add May 15th 2019 MTP trigger time
Rename tweak minigForkTime user facing name to consensus.forkMay2019Time

sickpig committed:
Fork helpers for May 15th 2019 protocol upgrade activation

Along with IsMay152019Next and IsMay152019Enabled, which will
be used to determine if the conditions for May 15th protocol
upgrade activation are met, we add also AreWeOnBCHChain and
AreWeOnSVChain which will return true/false according to which
chain bitcoind is set to follow.

sickpig committed:
Change nov 15th, 2018 forks helper to use height rather than time

sickpig committed:
Change default value of ctor tweaks to true.

sickpig committed:
Tune consensus rules in early boostrap phase.

Adapt consensus rules according to which chain we are following
and current block height in the early stages of execution.

On BCH ctor is turned on by default regardless of ctor tweak value
unless we are on regtest. On BSV we immediately increase the max
number of ops per script and max block size.

sickpig committed:
Activated Nov 15th, 2019 upgrade by block height.

This has been changed for block validation, block
construction and transaction admission to the mempool.

sickpig committed:
Update forkTime tweak validator to use may2019ActivationTime

sickpig committed:
Disable CDSV activation python test

Keep the file around as template for future op activation

sickpig committed:
Adapt ctor to use regtest only enableCanonicalTxOrder tweak

Simulate ctor activation leveraging the fact that we turn
enableCanonicalTxOrder into a regtest only tweak.

sickpig committed:
Removing dead code from mining_ctor test

now that ctor is enable by default there's no need to use mocktime
to enable ctor.

ptschip commented on src/validation/forks.h:
```diff
@@ -42,12 +42,19 @@ bool IsDAAEnabled(const Consensus::Params &consensusparams, const CBlockIndex *p
 
-/** Test if fork is active */
+/** Test if May 1th 2019 fork has active */
 bool IsMay152019Enabled(const Consensus::Params &consensusparams, const CBlockIndex *pindexTip);
```
May1th > May25th    

active -> activated

sickpig replied:
will fix

ptschip commented on src/init.cpp:
```diff
@@ -1235,23 +1235,18 @@ bool AppInit2(Config &config, thread_group &threadGroup)
-        }
+        maxScriptOps = SV_MAX_OPS_PER_SCRIPT;
+        excessiveBlockSize = SV_EXCESSIVE_BLOCK_SIZE;
+        settingsToUserAgentString();
```
shouldn't you set canonicaltxorder here to false here? I though I saw earlier you set the global tweak to true?

sickpig replied:
you're right will fix

ptschip commented on src/miner.cpp:
```diff
@@ -169,8 +169,8 @@ CTransactionRef BlockAssembler::coinbaseTx(const CScript &scriptPubKeyIn, int _n
-        IsNov152018Enabled(Params().GetConsensus(), chainActive.Tip()))
+    if (nCoinbaseSize < MIN_TX_SIZE &&
+        (AreWeOnBCHChain() && IsNov152018Activated(Params().GetConsensus(), chainActive.Tip())))
     {
```
for readability, AreWeOnBCHChain() could logically include the test for isnov152018acivated()?

sickpig replied:
no strong opinion, I think that we are going to use it also for future forks while moving from date based to height based activation. 

but I think you have a point, thou. In the future we could just  modifying `AreWeOnBCHChain()`.

will change.

ptschip commented on src/validation/validation.cpp:
```diff
@@ -2427,17 +2437,11 @@ bool ConnectBlock(const CBlock &block,
-            canonical = false;
-        }
+        canonical = true;
     }
```
looks like you accidentally took out the   else{ canonical = false}  ?

sickpig replied:
you are right again, even if we checkpointed the nov 15th fork we should give a chance to people to IBD with full checking enabled.

ptschip commented on src/validation/validation.cpp:
```diff
@@ -2724,29 +2728,23 @@ void UpdateTip(CBlockIndex *pindexNew)
-            enableCanonicalTxOrder = false;
-        }
+        enableCanonicalTxOrder = AreWeOnBCHChain();
     }
```
again you took out ht enablecanonical=false

sickpig replied:
ditto

sickpig committed:
Ctor is always disabled on BSV

sickpig committed:
Fix typo in a comment

sickpig committed:
Embed IsNov152018Activated() check on AreWeOnCBHChain()

sickpig committed:
Turns enableCanonicaTxOrder default back to false

sickpig committed:
Switch to DEFAULT_EXCESSIVE_BLOCK_SIZE only on BCH main chain.

dagurval commented on qa/pull-tester/rpc-tests.py:
```diff
@@ -236,7 +236,7 @@ def option_passed(option_without_dashes):
     'grapheneblocks',
     'thinblocks',
-    'checkdatasig_activation',
+    Disabled('checkdatasig_activation', "CDSV has been already succesfully activated, keep test around as a template for other OP activation"),
```
Keeping old stuff around is git's purpose

sickpig replied:
agreed, I was just following the same convention that we use for a bunch of other tests we disabled in the past. e.g. : 

```python
    Disabled('nov152018_forkactivation','Nov 2018 already activated'),
    Disabled('may152018_forkactivation_1','May 2018 already activated, use it as template to test future upgrade activation'),
    Disabled('may152018_forkactivation_2','May 2018 already activated, use it as template to test future upgrade activation'),
```

I have no problem to delete CDSV test since it is stored in git history in any case.

dagurval commented on src/globals.cpp:
```diff
@@ -336,7 +336,7 @@ CTweakRef<std::string> subverOverrideTweak("net.subversionOverride",
 CTweak<bool> enableCanonicalTxOrder("consensus.enableCanonicalTxOrder",
     "True if canonical transaction ordering is enabled.  Reflects the actual state so may be switched on or off by"
     " fork time flags and blockchain reorganizations.",
-    true);
```
The network default is on, why reverse this?

sickpig replied:
(we already settle this out band I'll add here my explanation for future reference)

Have it false by default makes sense case in case someone makes an IBD with `checklevel=4` and `checkpoints=0` (i.e. full validation also for historical blocks) in fact from genesis till Nov 2018 activation ctor should not be enforced. 

Doing the other way around would mean more convoluted code IMHO.

ptschip commented on src/validation/validation.cpp:
```diff
@@ -2748,7 +2748,10 @@ void UpdateTip(CBlockIndex *pindexNew)
+        if (chainParams.NetworkIDString() != "regtest")
+        {
+            excessiveBlockSize = DEFAULT_EXCESSIVE_BLOCK_SIZE;
+        }
```
I disagree with this...problem is if users on mainnet want to change their Excessive block size manually it won't stick after the next block is mined...I think there's a logic error somewhere in your new code changes.

dagurval replied:
Also this does not do what the commit message says it does

sickpig replied:
Agreed I should have split the `if` condition in 2, something like: 

```c++
    if (AreWeOnSVChain())
    {
        if (IsSv2018Activated(chainParams.GetConsensus(), chainActive.Tip()))
        {
            maxScriptOps = SV_MAX_OPS_PER_SCRIPT;
            excessiveBlockSize = SV_EXCESSIVE_BLOCK_SIZE;
        }
        else // if blockchain reorg we may need to back it out
        {
            maxScriptOps = MAX_OPS_PER_SCRIPT;
            excessiveBlockSize = DEFAULT_EXCESSIVE_BLOCK_SIZE;
        }
    }
```

current `else` branch was supposed to be execute only in the BSV case, on the contrary the current code execute the `else` branch also BCH branch which is a logic bug. 

fixing.

sickpig replied:
fixed ea81d7a3c7

dagurval commented on src/validation/forks.h:
```diff
@@ -21,6 +21,8 @@ class CBlockIndex;
 class CScript;
 class CTxMemPoolEntry;
 
+extern CChain chainActive;
```
This looks out of place. If this is due to having issues including `main.h`, then there is probably a deeper code coupling problem that needs to be solved.

sickpig replied:
This will be solved by passing param and current tip to the `AreWeOnBCHChain()` as you sugegest here https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1644#discussion_r272368068

sickpig replied:
done, see fa23e17

dagurval commented on src/validation/forks.cpp:
```diff
@@ -97,6 +96,12 @@ bool IsNov152018Activated(const Consensus::Params &consensusparams, const CBlock
 }
 
+bool AreWeOnBCHChain()
+{
```
Rather than overriding the parameters passed to `IsNov152018Activated`, `AreWeOnBCHChain` should take `consensusparams` and `tip` as arguments.

dagurval replied:
I find the name of this function to be misleading. It is checking if you're on BCH *AND* past Nov 2018 fork time. The need to comment "Nov 15th" several places elsewhere is a symptom of this.

How about putting it in namespace `bch` and calling it `IsNov2018Activated`? Usage would be something like:
```
const uint32_t cds_flag = bch::IsNov2018Activated(params, tip) ?
            SCRIPT_ENABLE_CHECKDATASIG :
            0;
```

sickpig replied:
> I find the name of this function to be misleading. It is checking if you're on BCH AND past Nov 2018 fork time.

Indeed. Originally the check for `IsNov2018Activated` wasn't included in the `AreWeOnBCHChain` function. 

On second thought I wonder if it would have been better to keep them separated. Will reconsider tomorrow. 

Having said I find the idea of using `bch::IsNov2018Activated(params, tip)` a good one. 

If I decide to maintain both condition in the same function I will implement your suggested changes.

gandrewstone replied:
Previously we used whether a fork time was configured to determine the SV or BCH chains.  But this method is going to be very confusing going forward, especially since the next fork date may not be known.  

At this point, we should add a tweak "svEnabled" (default = false) that chooses the fork rather than using these forktime tweaks.  This will be very understandable to users.  If a fork is being prepared on one chain, we then can create one or two fork time variables that define when the fork will happen.

As differences mount, we may need to move to a separate release.  But I don't think that that is needed for this release.

sickpig replied:
Agreed.

dagurval commented on src/chainparams.cpp:
```diff
@@ -139,7 +140,10 @@ class CMainParams : public CChainParams
+        consensus.nov2018Height = 556766;
+        consensus.sv2018Height = 556766;
+        // Wed, 15 May 2019 12:00:00 UTC hard fork activation time
+        consensus.may2019ActivationTime = 1557921600;
```
I verified that `1557921600` is indeed 15th may 2019 12pm UTC.

dagurval commented on src/validation/forks.cpp:
```diff
@@ -92,41 +82,57 @@ bool IsDAAEnabled(const Consensus::Params &consensusparams, const CBlockIndex *p
-
-bool IsSv2018Scheduled() { return miningSvForkTime.Value() != 0; }
-bool IsSv2018Enabled(const Consensus::Params &consensusparams, const CBlockIndex *pindexTip)
+bool IsMay152019Next(const Consensus::Params &consensusparams, const CBlockIndex *pindexTip)
```
Nit: There is only one fork in May, so maybe remove `15` for readability

sickpig replied:
good idea. 

I was just following the naming convention we used so far. 

no problem in changing it. 

let's see what other thinks.

sickpig replied:
@ptschip @gandrewstone what's your take? 

do you prefer `IsMay152019...` or `IsMay2019...`? 

@dagurval  have indeed a point the 15th is redundant info, It's not that we are going to have multiple upgrades in May (hopefully :smile:)

ptschip replied:
makes sense

sickpig replied:
changed in 66e065ca

dagurval commented on src/validation/forks.cpp:
```diff
@@ -92,41 +82,57 @@ bool IsDAAEnabled(const Consensus::Params &consensusparams, const CBlockIndex *p
+
+//* SV helpers/
+
+bool AreWeOnSVChain() { return miningSvForkTime.Value() != 0; }
```
Suggest `assert(miningForkTime == 0)` here and the reverse in `AreWeOnBCHChain`. Don't use the DbgAssert, it's good to die when trying to enable incompatible consensus rules.

sickpig replied:
`AreWeOnSVChain()` is meant to turn false if `miningForkTime != 0`, i.e. `miningSvForkTime == 0` rather than asserting i.e. crashing. 

The tweaks validator guarantee that if you can't set `miningForkTime` and `miningSvForkTime` both to something greater than `0`.

See line 232 and 256 of `src/unlimited.cpp`

So it seems impossible to me that `AreWeOnSVChain()` return true and at the same time we are going to apply BCH rule set.

dagurval commented on src/validation/validation.cpp:
```diff
@@ -2724,27 +2730,26 @@ void UpdateTip(CBlockIndex *pindexNew)
+        maxScriptOps = SV_MAX_OPS_PER_SCRIPT;
+        excessiveBlockSize = SV_EXCESSIVE_BLOCK_SIZE;
+    }
+    else
```
Nit: Ignoring ptschip's comment below about correctness, it's possible to simplify and remove the else clause here by doing:

```
maxScriptOps = MAX_OPS_PER_SCRIPT;
excessiveBlockSize = DEFAULT_EXCESSIVE_BLOCK_SIZE;

if (AreWeOnSVChain()) {

	// override with sv stuff
}
```

sickpig replied:
@ptschip was right there was a logic error in the prev code. In fact the EB change was meant to be executed only on the SV chain. see ea81d7a for the new code

sickpig committed:
EB was meant to be change back to 32MB only in case of a reorg in the BSV chain

As the code was stated before the Eb change could have been executed also
when on BCH chain (i.e. AreWeOnSVChain == false)

ptschip commented:
@sickpig I don't see any code in miner.cpp that will handle the switchover to schnorr as far as building blocks...in BlockAssemberl:TestForBlock i think there needs to be a check for whether this tx type is valid or not otherwise if there are still old tx types in the mempool after the fork, they may get added to a newly mined block and then the block will fail validation.

ptschip commented:
@sickpig I wonder if it makes better sense to just clear the mempool at forktime...

sickpig commented:
> @sickpig I don't see any code in miner.cpp that will handle the switchover to schnorr as far as building blocks...in BlockAssemberl:TestForBlock i think there needs to be a check for whether this tx type is valid or not otherwise if there are still old tx types in the mempool after the fork, they may get added to a newly mined block and then the block will fail validation.

@ptschip the PR is not meant to provide the activation code for schnorr nor for segwit recovery changes. 

This PR just introduce two helpers to deal with the fork activation (` IsMay152019Enabled` and `IsMay152019Next`), those have to be used to implement the activation code for all consensus changes contained in the upcoming upgrade.

The other thing this in this PR is the change to the code activation for Nov 2018 upgrade which now uses block height rather than MTP.

ptschip commented:
sure , makes sense

On 05/04/2019 1:10 a.m., Andrea Suisani wrote:
>
> *@sickpig* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/validation/forks.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1644#discussion_r272483239>:
>
> >  }
>
>  
>
> -
>
> -bool IsSv2018Scheduled() { return miningSvForkTime.Value() != 0; }
>
> -bool IsSv2018Enabled(const Consensus::Params &consensusparams, const CBlockIndex *pindexTip)
>
> +bool IsMay152019Next(const Consensus::Params &consensusparams, const CBlockIndex *pindexTip)
>
> @ptschip <https://github.com/ptschip> @gandrewstone
> <https://github.com/gandrewstone> what's your take?
>
> do you prefer |IsMay152019...| or |IsMay2019...|?
>
> @Dagur <https://github.com/Dagur> have indeed a point the 15th is
> redundant info, It's not that we are going to have multiple upgrades
> in May (hopefully 😄)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1644#discussion_r272483239>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0Obz0BbZSgysA12qqlNw3_8QVpHDks5vdwUEgaJpZM4cbhSZ>.
>

sickpig committed:
Rename May 2019 fork activation helpers

s/IsMay152019/IsMay2019/

sickpig committed:
Pass consensus params and current chaintip to AreWeOnBCHChain.

gandrewstone commented:
> @sickpig I don't see any code in miner.cpp that will handle the switchover to schnorr as far as building blocks...in BlockAssemberl:TestForBlock i think there needs to be a check for whether this tx type is valid or not otherwise if there are still old tx types in the mempool after the fork, they may get added to a newly mined block and then the block will fail validation.

most old tx will work, but you are right, an ECDSA sig of exactly 64 bytes will fail.  The easiest thing to do is once the last pre-fork block is found, we should take all the tx out of the mempool and enqueue them onto the tx admission queue.

ptschip commented on src/txadmission.cpp:
```diff
@@ -598,11 +598,13 @@ bool ParallelAcceptToMemoryPool(Snapshot &ss,
 
     const uint32_t cds_flag =
-        (AreWeOnBCHChain(chainparams.GetConsensus(), chainActive.Tip())) ? SCRIPT_ENABLE_CHECKDATASIG : 0;
+      (AreWeOnBCHChain(chainparams.GetConsensus(), chainActive.Tip())) ? SCRIPT_ENABLE_CHECKDATASIG : 0;
```
incorrect indentation

sickpig commented:
@gandrewstone an ECSDA of exactly 64 bit will fail because of the 100 bytes constraint.

> pre-fork block is found, we should take all the tx out of the mempool and enqueue them onto the tx admission queue.

agreed. But again this PR wasn't meant to provide any May 2019 activation code, just 2 helpers and switch the Nov 2018 one to be triggered by height rather than time.

sickpig commented:
@gandrewstone fixed the indentation problem by rebasing my local version of the branch on top of the current `dev`.

you could do it to via cli by adding to your repo mine via `git remoted add` using the URL based on the git+ssh protocol rather then https, i.e.

```bash
git remote add sickpig git@github.com:BitcoinUnlimited/BitcoinUnlimited.git
git checkout prepare-for-the-fork-2
git rebase ustream/dev
git push -f
```

gandrewstone commented on src/init.cpp:
```diff
@@ -1235,23 +1235,18 @@ bool AppInit2(Config &config, thread_group &threadGroup)
-        {
-            enableCanonicalTxOrder = true;
-        }
+        enableCanonicalTxOrder = true;
```
what if we are in IBD, prior to the Nov2018 fork?

gandrewstone commented on src/validation/forks.cpp:
```diff
@@ -92,41 +82,56 @@ bool IsDAAEnabled(const Consensus::Params &consensusparams, const CBlockIndex *p
-bool IsNov152018Enabled(const Consensus::Params &consensusparams, const CBlockIndex *pindexTip)
+bool IsNov152018Activated(const Consensus::Params &consensusparams, const int32_t nHeight)
+{
+    return nHeight >= consensusparams.nov2018Height;
```
determine SV vs BCH first, then use the appropriate height?

sickpig committed:
Add SV testnet activation height

sickpig committed:
Decouple IsNov152018Activated() from AreWeOnCBHChain()

This is basically a revert of 933057d0cd ("Embed IsNov152018Activated() check on AreWeOnCBHChain()).
AreWeOnBCHChain() was basically meant to be a quick
way to tell you if node is configured to follow BCH
fork. The aforementioned commit overloaded the function
with other meaning, namely keeping track if Nov 2018 BCH
network upgrade happened already.

While this bring some sort of read clarity to the code
it introduces some ambiguity the could lead error in the
use of the function so we decided to decouple IsNov152018Activated().

sickpig committed:
s/IsNov152018Activated/IsNov2018Activated/
